libwnbd/wnbd_ioctl.cpp

```diff
@@ -133,6 +137,245 @@ DWORD WnbdOpenAdapterCMDeviceInstance(PDEVINST DeviceInstance)
     return Status;
 }
 
+DWORD GetDrvDetailData(HDEVINFO DevHandle, PSP_DEVINFO_DATA DevInfoData,
+                       PSP_DRVINFO_DETAIL_DATA_A DrvDetailData)
+{
+    SP_DEVINSTALL_PARAMS_A InstallParams;
+    BOOL Cleanup = FALSE, Found = FALSE;
+    SP_DRVINFO_DATA_A DrvInfoData;
+    DrvInfoData.cbSize = sizeof DrvInfoData;
+    DWORD GLA = ERROR_SUCCESS;
+
+    InstallParams.cbSize = sizeof InstallParams;
+    if (!SetupDiGetDeviceInstallParamsA(DevHandle, DevInfoData, &InstallParams)) {
+        GLA = GetLastError();
+        goto exit;
+    }
+
+    InstallParams.FlagsEx |= DI_FLAGSEX_INSTALLEDDRIVER | DI_FLAGSEX_ALLOWEXCLUDEDDRVS;
+    if (!SetupDiSetDeviceInstallParamsA(DevHandle, DevInfoData, &InstallParams)) {
+        GLA = GetLastError();
+        goto exit;
+    }
+
+    if (!SetupDiBuildDriverInfoList(DevHandle, DevInfoData, SPDIT_CLASSDRIVER)) {
+        GLA = GetLastError();
+        goto exit;
+    }
+    Cleanup = TRUE;
+
+    for (DWORD index = 0;
+        SetupDiEnumDriverInfoA(DevHandle, DevInfoData, SPDIT_CLASSDRIVER, index, &DrvInfoData);
```

Comment:
I think we should move ``SetupDiEnumDriverInfoA`` inside the loop so that we can check the error code. If it's something other than ERROR_NO_MORE_ITEMS, we should log and propagate the error.

Same for ``SetupDiGetDriverInfoDetailA`` errors, I think we may want to log/propagate them.

Reply:
Yep, but we may fail to locate it due to some unrelated error, even if the driver exists. In that case, the error should be logged and propagated.

Reply:
> 
> 
> Yep, but we may fail to locate it due to some unrelated error, even if the driver exists. In that case, the error should be logged and propagated.

`the error will be propagated via line 181 and logged at line 188 , because Found will be false.`